README.md
# firehose-write-stream

A port from [kinesis-write-stream](https://github.com/voldern/kinesis-write-stream) for use with AWS Firehose.
Firehose writable stream that buffers up records

# Usage

Records written to the stream will buffer up until `highWaterMark` has
been reached, or the stream is closed, before writing to Firehose using
`putRecordBatch`.

## Failed items

Failed items will be retried up until `options.maxRetries` has been
reached. The initial timeout before the retry is set in
`options.retryTimeout` and it increases by the fibonnaci sequence.

## Logging

A [bunyan](https://www.npmjs.com/package/bunyan),
[winston](https://www.npmjs.com/package/winston) or similar logger
instance that have methods like `debug`, `error` and `info` may be
sent in as `options.logger` to the constructor.

# Example

```javascript
var FirehoseWritable = require('firehose-write-stream');

var stream = new FirehoseWritable(firehoseClient, 'streamName', {
  highWaterMark: 16,
  maxRetries: 3,
  retryTimeout: 100
});

inputStream.pipe(stream);
```

# API

See [api.md](api.md).

# License

MIT


CHANGELOG.md
# Change Log
## [1.0.0] - 2017-01-06
Ported [kinesis-write-stream](https://github.com/voldern/kinesis-write-stream) for use with AWS Firehose.
A big thanks to Espen Volden


test/firehose-write-stream.js
'use strict';

var chai = require('chai'),
    sinon = require('sinon'),
    sinonChai = require('sinon-chai'),
    streamArray = require('stream-array'),
    FirehoseWritable = require('../'),
    recordsFixture = require('./fixture/records'),
    successResponseFixture = require('./fixture/success-response'),
    errorResponseFixture = require('./fixture/failed-response'),
    writeFixture = require('./fixture/write-fixture');

chai.use(sinonChai);

var expect = chai.expect;

describe('FirehoseWritable', function() {
    beforeEach(function() {
        this.sinon = sinon.sandbox.create();

        this.client = {
            putRecordBatch: sinon.stub()
        };

        this.stream = new FirehoseWritable(this.client, 'streamName', {
            highWaterMark: 6
        });
    });

    afterEach(function() {
        this.sinon.restore();
    });

    describe('constructor', function() {
        it('should throw error on missing client', function() {
            expect(function() {
                new FirehoseWritable();
            }).to.Throw(Error, 'client is required');
        });

        it('should throw error on missing streamName', function() {
            expect(function() {
                new FirehoseWritable({});
            }).to.Throw(Error, 'streamName is required');
        });

        it('should throw error on highWaterMark above 500', function() {
            expect(function() {
                new FirehoseWritable({}, 'test', { highWaterMark: 501 });
            }).to.Throw(Error, 'Max highWaterMark is 500');
        });
    });

    describe('_write', function() {
        it('should write to firehose when stream is closed', function(done) {
            this.client.putRecordBatch.yields(null, successResponseFixture);

            this.stream.on('finish', function() {
                expect(this.client.putRecordBatch).to.have.been.calledOnce;

                expect(this.client.putRecordBatch).to.have.been.calledWith({
                    Records: writeFixture,
                    DeliveryStreamName: 'streamName'
                });

                done();
            }.bind(this));

            streamArray(recordsFixture)
                .pipe(this.stream);
        });

        it('should do nothing if there is nothing in the queue when the stream is closed', function(done) {
            this.client.putRecordBatch.yields(null, successResponseFixture);

            this.stream.on('finish', function() {
                expect(this.client.putRecordBatch).to.have.been.calledOnce;

                done();
            }.bind(this));

            for (var i = 0; i < 6; i++) {
                this.stream.write(recordsFixture);
            }

            this.stream.end();
        });

        it('should buffer records up to highWaterMark', function(done) {
            this.client.putRecordBatch.yields(null, successResponseFixture);

            for (var i = 0; i < 4; i++) {
                this.stream.write(recordsFixture[0]);
            }

            this.stream.write(recordsFixture[0], function() {
                expect(this.client.putRecordBatch).to.not.have.been.called;

                this.stream.write(recordsFixture[0], function() {
                    expect(this.client.putRecordBatch).to.have.been.calledOnce;

                    done();
                }.bind(this));
            }.bind(this));
        });

        it('should emit error on Firehose error', function(done) {
            this.client.putRecordBatch.yields('Fail');

            this.stream.on('error', function(err) {
                expect(err).to.eq('Fail');

                done();
            });

            this.stream.end({ foo: 'bar' });
        });

        it('should emit error on records errors', function(done) {
            this.client.putRecordBatch.yields(null, errorResponseFixture);

            this.stream.on('error', function(err) {
                expect(err).to.be.ok;

                done();
            });

            this.stream.end({ foo: 'bar' });
        });

        it('should retry failed records', function(done) {
            this.client.putRecordBatch.yields(null, errorResponseFixture);
            this.client.putRecordBatch.onCall(2).yields(null, successResponseFixture);

            this.stream.on('finish', function() {
                expect(this.client.putRecordBatch).to.have.been.calledThrice;

                expect(this.client.putRecordBatch.secondCall).to.have.been.calledWith({
                    Records: [
                        writeFixture[1],
                        writeFixture[3]
                    ],
                    DeliveryStreamName: 'streamName'
                });

                done();
            }.bind(this));

            streamArray(recordsFixture).pipe(this.stream);
        });
    });
});


test/fixture/write-fixture.js
'use strict';

var records = require('./records.json');

records = records.map(function(record) {
    return {
        Data: JSON.stringify(record)
    };
});

module.exports = records;


index.js
'use strict';

var util = require('util'),
    assert = require('assert'),
    retryFn = require('retry-fn'),
    FlushWritable = require('flushwritable');

util.inherits(FirehoseWritable, FlushWritable);

/**
 * A simple stream wrapper around Firehose putRecordBatch
 *
 * @param {AWS.Firehose} client AWS Firehose client
 * @param {string} streamName Firehose stream name
 * @param {Object} options Options
 * @param {number} [options.highWaterMark=16] Number of items to buffer before writing.
 * Also the size of the underlying stream buffer.
 * @param {Number} [options.maxRetries=3] How many times to retry failed data before
 * rasing an error
 * @param {Number} [options.retryTimeout=100] How long to wait initially before
 * retrying failed records
 * @param {Object} [options.logger] Instance of a logger like bunyan or winston
 * @param {String} [options.delimiter] String separator between Firehose record
 */
function FirehoseWritable(client, streamName, options) {
    assert(client, 'client is required');
    assert(streamName, 'streamName is required');

    options = options || {};
    options.objectMode = true;

    FlushWritable.call(this, options);

    this.client = client;
    this.streamName = streamName;
    this.logger = options.logger || null;
    this.highWaterMark = options.highWaterMark || 16;
    this.maxRetries = options.maxRetries || 3;
    this.delimiter = options.delimiter || '';

    // The maximum number of items that can be written to Firehose in
    // one request is 500
    assert(this.highWaterMark <= 500, 'Max highWaterMark is 500');

    this.queue = [];
}

/**
 * Transform record into one accepted by Firehose
 *
 * @private
 * @param {Object} record
 * @return {Object}
 */
FirehoseWritable.prototype.transformRecord = function transformRecord(record) {
    return {
        Data: JSON.stringify(record) + this.delimiter
    };
};

/**
 * Write records in queue to Firehose
 *
 * @private
 * @param {Function} callback
 */
FirehoseWritable.prototype.writeRecords = function writeRecords(callback) {
    if (this.logger) {
        this.logger.debug('Writing %d records to Firehose', this.queue.length);
    }

    var records = this.queue.map(this.transformRecord.bind(this));

    this.client.putRecordBatch({
        Records: records,
        DeliveryStreamName: this.streamName
    }, function(err, response) {
        if (err) {
            return callback(err);
        }

        if (this.logger) {
            this.logger.info('Wrote %d records to Firehose',
                records.length - response.FailedPutCount);
        }

        if (response.FailedPutCount !== 0) {
            if (this.logger) {
                this.logger.warn('Failed writing %d records to Firehose',
                    response.FailedPutCount);
            }

            var failedRecords = [];

            response.RequestResponses.forEach(function(record, index) {
                if (record.ErrorCode) {
                    if (this.logger) {
                        this.logger.warn('Failed record with message: %s', record.ErrorMessage);
                    }

                    failedRecords.push(this.queue[index]);
                }
            }.bind(this));

            this.queue = failedRecords;

            return callback(new Error('Failed to write ' + failedRecords.length + ' records'));
        }

        this.queue = [];

        callback();
    }.bind(this));
};

/**
 * Flush method needed by the underlying stream implementation
 *
 * @private
 * @param {Function} callback
 * @return {undefined}
 */
FirehoseWritable.prototype._flush = function _flush(callback) {
    if (this.queue.length === 0) {
        return callback();
    }

    var retry = retryFn.bind(null, {
        retries: this.maxRetries,
        timeout: retryFn.fib(this.retryTimeout)
    });

    retry(this.writeRecords.bind(this), callback);
};

/**
 * Write method needed by the underlying stream implementation
 *
 * @private
 * @param {Object} record
 * @param {string} enc
 * @param {Function} callback
 * @returns {undefined}
 */
FirehoseWritable.prototype._write = function _write(record, enc, callback) {
    if (this.logger) {
        this.logger.debug('Adding to Firehose queue', { record: record });
    }

    this.queue.push(record);

    if (this.queue.length >= this.highWaterMark) {
        return this._flush(callback);
    }

    callback();
};

module.exports = FirehoseWritable;
